### DISCUSSION TITLE
Frontend conventions/tooling

### DISCUSSION THREAD

**Discussion by notjosh** [slskd/slskd#1076] [Ideas]
Hiya,

I've been running a local copy of the code, and fixing a few bugs + performance issues here and there (mostly related to #317).

Today I finally got around to separating some of them to make PRs, but realised my editor had format-on-save'd and changed a lot. When I started untangling some of it, I realised the frontend's linting is fairly unrestrictive so there's a lot of mixed conventions.

Best to avoid holy wars in OSS projects (tabs! spaces! argh!), so how would you feel about pulling in a couple of packages to take out the guesswork?

My proposal would be to introduce [Prettier](https://prettier.io/) + a common ESLint preset (i.e. [`eslint-config-standard`](https://github.com/standard/eslint-config-standard)) to enforce consistency of most common parts.

I'm happy to put up a PR if you're interested in seeing how it looks, and we can iterate on any conventions/exceptions you want to define?

Don't have to go into them too much here, but a couple of other thoughts/questions:

- [`create-react-app`](https://github.com/facebook/create-react-app) (and by extension [`craco`](https://github.com/dilanx/craco)) seems to be mooostly abandoned, so some of the tooling with vulnerabilities aren't getting patched. In other projects I've replaced it fairly painlessly with [Vite](https://vitejs.dev/). Somethin' to think about.
- it'd be rad to have TypeScript, which both Craco & Vite support fairly painlessly. Any major reason not to?

**Comment by jpdillingham** [comment]
All of that sounds great, thanks!  The frontend is a bit of a growing concern because it changes infrequently these days and all of the tooling is based on the original Soulseek.NET example code I threw together over 5 years ago.

I'm open to pretty much any changes at this point.  I would eventually like to rewrite the UI using a better supported design system like Ant Design or Mantine.

I built the app from the ground up to support "drop-in" frontends but haven't fleshed out how that would actually work during builds and at runtime; maybe I should do that and it would ease a transition to a "v2" of the UI

**Comment by notjosh** [comment]
> all of the tooling is based on the original Soulseek.NET example code I threw together over 5 years ago.

Ah :) that makes sense. Lemme throw a couple of PRs up to get things moving in a more modern direction.

> I built the app from the ground up to support "drop-in" frontends but haven't fleshed out how that would actually work during builds and at runtime; maybe I should do that and it would ease a transition to a "v2" of the UI

Does that require much from the frontend side, or is that all mostly .net tooling? I haven't touched .net much in farrrr too long, so I'm way out of the loop on that.

I figured the frontend was decoupled, and it just hit the backend as just an API, so it could (theoretically!) run/exist entirely separately.

**Comment by jpdillingham** [comment]
The backend will serve anything in the `wwwroot` folder, so replacing the frontend is really as simple as dropping static files from a SPA build in.  This happens during the build today, and to support more than 1 frontend natively I'd need to figure out how to build and consolidate them and allow users to choose which they'd like to use.  There's a setting to choose a different folder for static files so technically there's a path but it's not very user friendly.

This is probably as simple as adding subfolders under wwwroot and adding a user-friendly way to swap on the fly.

**Comment by notjosh** [comment]
Okie doke, I've split the work into 3 PRs - the work is stacked, so 2 of the PRs are on my fork to keep the diffs sensible.

- code style: https://github.com/slskd/slskd/pull/1080 - this is a bit of a monster, as it touches most files in the project. I echo your sentiment in https://github.com/slskd/slskd/pull/482 that it's not a great PR, but the end result is (probably?) worth it to have cleaner diffs/contributions in future
- migrate Craco to Vite: https://github.com/notjosh/slskd/pull/1 - we don't _need_ this, but it simplifies introducing TypeScript later. It makes builds a lil faster too
- introduce TypeScript: https://github.com/notjosh/slskd/pull/2 - this is a large PR, as I wanted to get an idea of the destination. But it can be introduced gradually

Consider them more as drafts than final contributions - let's focus on if you/we think if the proposals introduce value, and _then_ we can figure how to break up/clean up the PRs to land them safely.

**Reply by jpdillingham** [reply]
Sorry for the delay, 0.20.0 went out and we're good to go on this.  I merged the first one and will go look for linting errors this evening.
